forum/qa-plugin/report-reason/q2apro-flag-reasons-page.php

```diff
@@ -0,0 +1,209 @@
+<?php
+
+class q2apro_flag_reasons_page extends q2apro_flag_reasons_validation {
+
+    private $directory;
+    private $urlToRoot;
+
+    protected $CUSTOM_REPORT_REASON_ID = 0;
+    protected $_reportReasonRequestJSON = null;
+
+    public function __construct() {
+        $REASON_LIST_KEYS = array_keys(qa_lang('q2apro_flagreasons_lang/REASON_LIST'));
+        $this->CUSTOM_REPORT_REASON_ID = end($REASON_LIST_KEYS);
+    }
+
+    public function load_module($directory, $urlToRoot) {
+        $this->directory = $directory;
+        $this->urlToRoot = $urlToRoot;
+    }
+
+    public function suggest_requests() {
+        return [
+            [
+                'title'   => 'Report flag', // title of page
+                'request' => 'report-flag', // request name
+                'nav'     => 'M', // 'M'=main, 'F'=footer, 'B'=before main, 'O'=opposite main, null=none
+            ],
+        ];
+    }
+
+    public function match_request($request) {
+        return 'report-flag' === $request;
+    }
+
+    public function process_request() {
+        $this->_reportReasonRequestJSON = file_get_contents('php://input');
+
+        if (!qa_is_logged_in()) {
+            $this->handleReportErrorAndExit('USER_LOGGED_OUT');
+        }
+
+        $flagData = $this->parseFlagData();
+
+        $questionId = (int) $flagData['questionId'];
+        $postId     = (int) $flagData['postId'];
+        $postType   = $flagData['postType'];
+        $reasonId   = (int) $flagData['reasonId'];
+
+        $notice = empty($flagData['notice']) ? null : trim($flagData['notice']);
+        $userId = qa_get_logged_in_userid();
+
+        require_once QA_INCLUDE_DIR . 'app/votes.php';
+        require_once QA_INCLUDE_DIR . 'app/posts.php';
+        require_once QA_INCLUDE_DIR . 'pages/question-view.php';
+
+        $processingFlagReasonError = $this->processFlag($postType, $userId, $postId, $questionId, $reasonId, $notice);
+
+        $reply = $processingFlagReasonError ?
+            $this->wrapFlagReasonError($processingFlagReasonError) :
+            ['newFlags' => $this->wrapOutput(q2apro_count_postflags_output($postId), $postType, $postId)];
+
+        echo json_encode($reply);
+    }
+
+    private function processFlag($postType, ...$flagParams) {
+        $processingFlagReasonError = '';
+
+        switch ($postType) {
+            case 'q': {
+                $processingFlagReasonError = call_user_func_array([$this, 'processFlagToQuestion'], $flagParams);
+                break;
+            }
```

Comment:
zapisywanie w klamrach konkretnego case też jest raczej niespotykane, nawet nie wiedziałem że taki zapis zadziała prawdę mówiąc

Reply:
Nie wiem jaką konwencję stosuje się w PHP, ale stosuję zasadę, że jeśli blok zawiera więcej niż jedną instrukcję, to należy go objąć w klamry, bo inaczej wykona tylko pierwszą z nich (czyli w tym przypadku `break;` już by się nie wykonał). Czy to nie ma racji bytu w przypadku `switch` w PHP?

Reply:
Nie, nie spotkałem takiego zapisu nawet. Tutaj również PSR-2 to określa: https://www.php-fig.org/psr/psr-2/#52-switch-case
A w JS się stosuje klamry do każdego case w switchu? Wydaje mi się, że też się tego nie robi, ale może zależy od preferencji.

Reply:
Hmm, okazuje się, że w JS `case`y w `switch`u można zapisywać [bez klamer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch#Syntax). Natomiast, gdy chce się stworzyć więcej niż jedną zmienną **o tej samej nazwie** słówkami `let`/`const`, użycie klamer 
w `case`ach [pozwala stworzyć zasięg blokowy](https://stackoverflow.com/a/42481784/4983840) (to już logiczne) zapobiegając rzuceniu błędem spowodowanym ponowną deklaracją zmiennej o tej samej nazwie.

Ja akurat z przyzwyczajenia zapisuje klamry dla wszystkich instrukcji blokowych. :) Jeśli w PHP to jest złą praktyką, to usunę klamry.

Reply:
Ciężko mi powiedzieć czy to zła praktyka, ja po prostu pierwszy raz zobaczyłem taki zapis w użyciu szczerze mówiąc :) PSR-2 o takim nie mówi, w innych tutorialach ani kodzie jakoś też nie spotkałem. Nie do końca też wyobrażam sobie jak wyglądało będzie użycie klamer, gdy wiele przypadków case ma pasować do jednego bloku kodu lub coś do niego dodawać, może będzie tak samo że dociągnie do break, ale wtedy te klamry powodowałyby raczej mylący zapis

Reply:
to jest do poprawy